staging/src/k8s.io/kubectl/pkg/cmd/apply/applyset.go

```diff
@@ -117,12 +190,223 @@ func (a *ApplySet) addLabels(objects []*resource.Info) error {
 	return nil
 }
 
-func (p *ApplySetParentRef) IsNamespaced() bool {
-	return p.RESTMapping.Scope.Name() == meta.RESTScopeNameNamespace
+// MarkObjectVisited keeps track of the applied objects, so that we can update the list of in-use namespaces
+// and resouce kinds, and so that we can prune objects not applied.
+func (a *ApplySet) MarkObjectVisited(resource *meta.RESTMapping, namespace string) {
+	a.updatedResources[resource.Resource] = resource
+	if namespace != "" {
+		a.updatedNamespaces.Insert(namespace)
+	}
+}
+
+func (a *ApplySet) FetchParent() error {
+	helper := resource.NewHelper(a.client, a.parent.RESTMapping)
+	obj, err := helper.Get(a.parent.Namespace, a.parent.Name)
+	if errors.IsNotFound(err) {
+		return nil
+	} else if err != nil {
+		return fmt.Errorf("failed to fetch ApplySet parent object %q from server: %w", a.ParentRef(), err)
+	} else if obj == nil {
+		return fmt.Errorf("failed to fetch ApplySet parent object %q from server", a.ParentRef())
+	}
+
+	labels, annotations := getLabelsAndAnnotations(obj)
+
+	toolAnnotation, hasToolAnno := annotations[ApplySetToolingAnnotation]
+	if !hasToolAnno {
+		return fmt.Errorf("ApplySet parent object %q already exists and is missing required annotation %q", a.ParentRef(), ApplySetToolingAnnotation)
+	}
+	if managedBy := toolingBaseName(toolAnnotation); managedBy != a.toolingID.name {
+		return fmt.Errorf("ApplySet parent object %q already exists and is managed by tooling %s instead of %s", a.ParentRef(), managedBy, a.toolingID.name)
+	}
+
+	idLabel, hasIDLabel := labels[ApplySetParentIDLabel]
+	if !hasIDLabel {
+		return fmt.Errorf("ApplySet parent object %q exists and does not have required label %s", a.ParentRef(), ApplySetParentIDLabel)
+	}
+	if idLabel != a.ID() {
+		return fmt.Errorf("ApplySet parent object %q exists and has incorrect value for label %q (got: %s, want: %s)", a.ParentRef(), ApplySetParentIDLabel, idLabel, a.ID())
+	}
+
+	if a.currentResources, err = parseResourcesAnnotation(annotations, a.restMapper); err != nil {
+		// TODO: handle GVRs for now-deleted CRDs
+		return fmt.Errorf("parsing ApplySet annotation on %q: %w", a.ParentRef(), err)
+	}
+	a.currentNamespaces = parseNamespacesAnnotation(annotations)
+	if a.parent.IsNamespaced() {
+		a.currentNamespaces.Insert(a.parent.Namespace)
+	}
+	return nil
+}
+
+func getLabelsAndAnnotations(obj runtime.Object) (map[string]string, map[string]string) {
+	accessor := meta.NewAccessor()
+	annotations := make(map[string]string)
+	labels := make(map[string]string)
+
+	if data, err := accessor.Annotations(obj); err == nil {
```

Comment:
I was wondering why this would return an error.  If we switch to `meta.Accessor(obj)` then if that returns an error I think we should return the error, and then `Labels()` and `Annotations()` will not return an error.